# Combining Data

Practice combining data from two different data sets. In the same folder as this Jupyter notebook, there are two csv files:
* rural_population_percent.csv
* electricity_access_percent.csv

They both come from the World Bank Indicators data. 
* https://data.worldbank.org/indicator/SP.RUR.TOTL.ZS
* https://data.worldbank.org/indicator/EG.ELC.ACCS.ZS

The rural populaton data represents the percent of a country's population that is rural over time. The electricity access data shows the percentage of people with access to electricity.

In this exercise, you will combine these two data sets together into one pandas data frame.

# Exercise 1

Combine the two data sets using the [pandas concat method](https://pandas.pydata.org/pandas-docs/stable/merging.html). In other words, find the union of the two data sets.

In [1]:
# TODO: import the pandas library
import pandas as pd
# TODO: read in each csv file into a separate variable
# HINT: remember from the Extract material that these csv file have some formatting issues
# HINT: The file paths are 'rural_population_percent.csv' and 'electricity_access_percent.csv'
df_rural = pd.read_csv('rural_population_percent.csv', skiprows=4)
df_electricity = pd.read_csv('electricity_access_percent.csv', skiprows=4)

# TODO: remove the Unnamed: 62 column from each data set
df_rural=df_rural.drop(labels='Unnamed: 62', axis=1)
df_electricity .drop(columns=['Unnamed: 62'], inplace=True)
# TODO: combine the two data sets together using the concat method
# In other words, all of the rows of df_rural will come first
# followed by all the rows in df_electricity. This is possible to do
# because they both have the same column names.
df=pd.concat([df_rural, df_electricity], axis=1)

In [2]:
df.head()

Country Name Country Code                            Indicator Name  \
0        Aruba          ABW  Rural population (% of total population)   
1  Afghanistan          AFG  Rural population (% of total population)   
2       Angola          AGO  Rural population (% of total population)   
3      Albania          ALB  Rural population (% of total population)   
4      Andorra          AND  Rural population (% of total population)   

   Indicator Code    1960    1961    1962    1963    1964    1965  ...   \
0  SP.RUR.TOTL.ZS  49.224  49.239  49.254  49.270  49.285  49.300  ...    
1  SP.RUR.TOTL.ZS  91.779  91.492  91.195  90.890  90.574  90.250  ...    
2  SP.RUR.TOTL.ZS  89.565  89.202  88.796  88.376  87.942  87.496  ...    
3  SP.RUR.TOTL.ZS  69.295  69.057  68.985  68.914  68.842  68.770  ...    
4  SP.RUR.TOTL.ZS  41.550  39.017  36.538  34.128  31.795  29.555  ...    

         2008        2009        2010        2011        2012        2013  \
0   93.086166   93.354546   93.356292   93.942375   94.255814   94.578262   
1   42.400000   44.854885   42.700000   43.222019   69.100000   67.259552   
2   31.268013   32.382469   33.514950   34.600000   35.821964   36.990490   
3  100.000000  100.000000  100.000000  100.000000  100.000000  100.000000   
4  100.000000  100.000000  100.000000  100.000000  100.000000  100.000000   

         2014        2015        2016  2017  
0   94.906723   95.238182   95.570145   NaN  
1   89.500000   71.500000   84.137138   NaN  
2   32.000000   42.000000   40.520607   NaN  
3  100.000000  100.000000  100.000000   NaN  
4  100.000000  100.000000  100.000000   NaN  

[5 rows x 124 columns]

# Exercise 2 (Challenge)

This exercise is more challenging.

Combine the two datas in the csv file together so that the output looks like the following:

|Country Name|Country Code|Year|Rural_Value|Electricity_Value|
|--|--|--|--|--|
|Aruba|ABW|1960|49.224|49.239|
... etc.

Order the results in the dataframe by country and then by year

Here are a few pandas methods that should be helpful:
* [melt](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html)
* [drop](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)
* [merge](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.DataFrame.merge.html)
* [sort_values](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)

HINT: You can use country name, country code, and the year as common keys between the data sets

In [3]:
# TODO: merge the data sets together according to the instructions. First, use the 
# melt method to change the formatting of each data frame so that it looks like this:
# Country Name, Country Code, Year, Rural Value
# Country Name, Country Code, Year, Electricity Value
df_rural = pd.read_csv('rural_population_percent.csv', skiprows=4)
df_electricity = pd.read_csv('electricity_access_percent.csv', skiprows=4)

df_rural=df_rural.drop(labels='Unnamed: 62', axis=1)
df_electricity .drop(columns=['Unnamed: 62'], inplace=True)

df_rural_melt=pd.melt(df_rural, id_vars=['Country Name','Country Code','Indicator Name','Indicator Code'], 
                 var_name='Year',
                 value_name='rural_value')

df_electricity_melt=pd.melt(df_electricity, id_vars=['Country Name','Country Code','Indicator Name','Indicator Code'], 
                       var_name='Year',
                 value_name='electricity_value')
# TODO: drop any columns from the data frames that aren't needed
df_rural_melt=df_rural_melt.drop(labels=['Indicator Name','Indicator Code'], axis=1)
df_electricity_melt=df_electricity_melt.drop(labels=['Indicator Name','Indicator Code'], axis=1)
# TODO: merge the data frames together based on their common columns

# in this case, the common columns are Country Name, Country Code, and Year
df_combined=df_rural_melt.merge( df_electricity_melt, on=['Country Name','Country Code','Year'], how='outer')
# TODO: sort the results by country and then by year

df_combined =df_combined.sort_values(by=['Country Name', 'Year'], ascending=[True, True])

In [4]:
df_combined.head()

Country Name Country Code  Year  rural_value  electricity_value
1     Afghanistan          AFG  1960       91.779                NaN
265   Afghanistan          AFG  1961       91.492                NaN
529   Afghanistan          AFG  1962       91.195                NaN
793   Afghanistan          AFG  1963       90.890                NaN
1057  Afghanistan          AFG  1964       90.574                NaN